### ISSUE TITLE
Integrate npm

### ISSUE THREAD

**Issue by whit537** [gratipay/gratipay.com#4148]
-------------------------------

✈️  **This is the flight deck for the [Integrate npm](https://github.com/orgs/gratipay/projects/5?fullscreen=true.) project.** ✈️ 

-------------------------------

Current open-source crowdfunding options (Kickstarter, Patreon, Gratipay, OpenCollective, etc.) are consumer-grade. Our hunch is that a business-grade product with better aggregation can better serve the companies that want to pay for open source, because companies use hundreds or thousands of open source packages, not just a few.

Picking up from https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-255122149 and https://github.com/gratipay/inside.gratipay.com/issues/852#issuecomment-255098337 ...

For wider context see:

- [The fundamental idea is to get a viral loop going again like we had under Gittipay 1.0.](https://github.com/gratipay/inside.gratipay.com/issues/637#issuecomment-263431528)
- [The burn I feel](http://gratipay.slackarchive.io/gratipay/-/1480699124.000841/1480708926.000944/1480708501000910/)

JavaScript is the most popular language in open source and npm is the most popular package manager for JavaScript. A good first concrete step towards helping companies pay for open source (#4135), therefore, will be to add the ability to pay for any package on npm. Once we have npm deployed, we will have enough experience to inform a partnership with Libraries.io for the rest of the package managers.

### Target

~~Our goal is to announce this feature in my [lightning talk](https://github.com/whit537/openorg/issues/6) on Thursday, October 26 at Red Hat's All Things Open conference (https://github.com/gratipay/inside.gratipay.com/issues/757).~~

Our goal is to incrementally improve this feature throughout the first half of 2017, with an eye towards [OSCON](https://github.com/gratipay/inside.gratipay.com/issues/948) and [$ustain](https://github.com/gratipay/inside.gratipay.com/issues/920) in May.

### Package names to test with

From https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-262672635:

http://localhost:8537/on/npm/async/
http://localhost:8537/on/npm/iframe-resizer/
http://localhost:8537/on/npm/mongoose/
http://localhost:8537/on/npm/nodemon/
http://localhost:8537/on/npm/react/
http://localhost:8537/on/npm/react-helmet/
http://localhost:8537/on/npm/react-modal/
http://localhost:8537/on/npm/react-redux/
http://localhost:8537/on/npm/react-router/
http://localhost:8537/on/npm/react-router-redux/
http://localhost:8537/on/npm/redux/
http://localhost:8537/on/npm/redux-thunk/
http://localhost:8537/on/npm/webpack/

#### Prerequisites

- [x] Land the Decouple milestone—https://github.com/gratipay/inside.gratipay.com/issues/432
- [ ] Finish Project Basics—https://github.com/gratipay/gratipay.com/milestone/18

#### Checkpoint 1: Inert `/on/npm/foo/` Pages

- [x] wrap marky-markdown (#4154)
- [x] load up npm (#4153, #4158)
- [x] load up npm readmes (#4159, #4160, #4161, #4162, #4164, #4176, #4179, #4180, #4181, #4182)
- [x] stop storing readmes after all (#4211)
- [x] add inert (can't actually give to them) `/on/npm/foo` pages (#4212)

#### Checkpoint 2: Giving to Packages

- [ ] link packages to teams (#4155)
- [ ] opt in or out from `/on/npm/foo/` pages via email verification
- [ ] add payment widget on `/on/npm/foo/` pages

#### Checkpoint 3: Easy Sign-up

- [ ] bulk manage payments from `/on/npm/` pages

#### Nice to Have

- [x] move search into `{% content %}` (leaving social network jump in sidebar)
- [ ] merge current result types together (~user via username, ~user via profile statement)
- [ ] add Teams to local index for inclusion in results
- [ ] mix npm packages into search (#4147)
- [ ] pledging to as-yet-unclaimed packages
- [ ] load up npm incrementally (reticket from #4153)
- [ ] deal intelligently with unpublished packages (reticket from https://github.com/gratipay/gratipay.com/issues/4148#issuecomment-256525647)
- [ ] we seem to get two H1s on https://www.npmjs.com/package/resize-img
- [ ] syntax highlight

#### Promotion

- [ ] update the "★ announcement ★" link (#4149)
- [ ] Publish a blog post.
- [ ] Promote the heck out of it!

---

✈️  **This is the flight deck for the [Integrate npm](https://github.com/orgs/gratipay/projects/5?fullscreen=true.) project.** ✈️

**Comment by whit537** [comment]
@mattbk You wanna draft a blog post?

**Comment by whit537** [comment]
@aandis How about I start hacking on this today and tomorrow, and I will leave notes for you over the weekend based on how far I get. Ya?

**Comment by whit537** [comment]
I've started a task list in the ticket description.

**Comment by aandis** [comment]
> @aandis How about I start hacking on this today and tomorrow, and I will leave notes for you over the weekend based on how far I get. Ya?

Yep. That works. A quickstart on https://github.com/gratipay/gratipay.com/pull/4147 will help in implementing the exact functionality you have in mind.

**Comment by whit537** [comment]
@aandis What's your availability? I was assuming Saturday and Sunday would be the time you could work on this but maybe I'm wrong?

**Comment by aandis** [comment]
@whit537 I am free over weekend, yes. I'll be keeping a close eye on this ticket for the next couple of days.

**Comment by whit537** [comment]
I've reached out to npm, Inc. in private email to let them know about our plans and invite them to give feedback and/or cross-promote with us next week.

**Comment by whit537** [comment]
Is there a rate limit on npm registry API requests? Can it be up'd with authentication? I find no docs on this.

**Comment by whit537** [comment]
We should think about possibly merging existing teams into these.

**Comment by whit537** [comment]
> The code that npm uses to talk to the registry.

https://github.com/npm/npm-registry-client

**Comment by aandis** [comment]
http://nipstr.com

**Comment by aandis** [comment]
@whit537 I am not seeing any official well documented api for searching packages over npm. Best I could find is https://github.com/wires/npm-keywordsearch through [this](http://stackoverflow.com/questions/13657140/how-to-get-all-npm-packages-that-match-a-particular-keyword-in-json-format).

**Comment by whit537** [comment]
Hashed this out a bit with @kaguillera. The problem we're solving is companies funding open source software. That's a market. The two sides are companies (buyers) and projects (sellers). We need to make it as easy as possible for companies to give to open source software; that's #4135. This is the other side, making it as easy as possible for developers to receive payments for open source.

### `/on/npm/`

One thing we realized is that a single developer will have multiple projects. "Easy as possible" means letting them configure multiple projects at once. Doug Wilson is maintaining [103 packages](https://www.npmjs.com/~dougwilson). We can't make him go through 103 separate verification workflows. He needs one worfklow.

We're thinking that that workflow should live at `/on/npm/`.

-**anon**—Sign in to configure payments for your npm packages.
-**auth**—Verify the email you use on npm to configure payments your npm packages.
-**no packages**—We didn't find any npm packages with alice@example.com as an author or maintainer.
-**packages**—We found 103 npm packages with alice@example.com as an author or maintainer.

We should account for the case where another ~user has already claimed the package.

Added to todo. ^^^ Bring this over to a PR when the time comes.

### Going all in? 

The other thing we realized is that the 400,000 packages on npm dwarf the 200 projects currently on Gratipay. We should consider redoing the homepage to contain a search box (for givers and the curious) and a link to `/on/npm/` (for developers ready to start receiving payments). We don't need to list "our Teams" on the homepage. With this change, "our Teams" instantly explodes to the size of npm.

**Comment by whit537** [comment]
Oh! And we should bring back pledging: givers should be able to set up payment instructions to projects before the projects opt-in. This is where opt-out starts to matter, because it prevents pledging.

**Comment by whit537** [comment]
We're getting back to the same dynamic as Gittipay 1.0, but with a focus on packages on open source software repositories (&rarr; Teams [Projects]) rather than users of social media networks (&rarr; ~users).

**Comment by whit537** [comment]
Also, it seems unlikely that we'll get this all done before next Thursday. Let's do what we can ... and have fun doing it! :-)

**Comment by whit537** [comment]
Alright, @aandis, I've stubbed out a few PRs and updated the todo in the description here. Hopefully that gives you something to go on before I get online tomorrow!

!m @aandis

**Comment by whit537** [comment]
@aandis Can we use this thread to communicate and coordinate about what specifically we're each working on? I'm seeing work from you on #4154 and #4155 (!m @aandis). What are you working on now? Do you think the todo on this ticket is right and in the right order or do we need to modify it?

I'm going to start at the top of the todo with #4117 and then move down to join you on the actual coding bits. :-)

Let's do this! 💃  🌻

**Comment by aandis** [comment]
@whit537 I am gonna wrap up #4154 and then work on #4153 and #4155 to make #4151 possible. I think  we should move pledging out of #4155 into it's own ticket and focus on mvp to be able to give to a package. Pledging is a nice to have feature but only when the groundwork is laid.

**Comment by whit537** [comment]
@aandis Cool. Strictly speaking, #4155 isn't necessary for #4151. In other words, we could have `/on/npm/foo/` pages without any possibility of giving to them. Is it worth batching #4154 #4153 and #4151 as our first checkpoint, and tackle #4155 after that first checkpoint is deployed?

**Comment by whit537** [comment]
{#4154, #4153, #4151} feels like something we could potentially get deployed before the end of the day today. Deployment itself will take some effort because of the need to spin up a worker dyno—something we haven't done before—to handle the continual npm syncing. Sure, inert `/on/npm/foo/` pages won't be much to brag about, but it'll at least give us a feeling of win. :)

Thoughts?

**Comment by aandis** [comment]
Sounds good. :)

**Comment by whit537** [comment]
@aandis Cool. I've updated the todo.

**Comment by whit537** [comment]
Alright, sooooooo ... it's emerging that the Decouple milestone, which should really get done before this (so we are free to accept all of the new Teams &rarr; projects we're hoping to get from this), is going to have some code implications. In particular we really ought to reflect a couple name changes in our codebase to avoid drift and tech debt.

- Teams &rarr; projects
- members &rarr; collaborators

I guess that's just a heads-up that we're looking at some potentially serious rebases of the work under here once the dust settles under https://github.com/gratipay/inside.gratipay.com/issues/432.

**Comment by mattbk** [comment]
> once the dust settles under gratipay/inside.gratipay.com#432

This will be a happy day.

**Comment by whit537** [comment]
Maybe Tuesday? 😳

**Comment by whit537** [comment]
> I guess that's just a heads-up that we're looking at some potentially serious rebases of the work under here once the dust settles under gratipay/inside.gratipay.com#432.

Scratch that, @aandis. It doesn't make sense to block this on the full codebase renames on gratipay/inside.gratipay.com#432. We can still get the important stuff done (Terms of Service update, in particular), without renaming tables.

**Comment by whit537** [comment]
@aandis Heads up that I'm going to be offline today. I've been going pretty hard and need a break before I fly out tomorrow for https://github.com/gratipay/inside.gratipay.com/issues/757. Whatever you get done under this will be awesome! Tomorrow afternoon/evening I should be able to pick up where you leave off.

Also, I think it may be better to focus on bulk sign-up on `/on/npm/` instead of adding npm packages to search. The former is of more value to developers, which is who we're aiming at here. The latter is for companies/givers, which is less of the focus this week.

**Comment by whit537** [comment]
!m @aandis 💃

**Comment by aandis** [comment]
@whit537 I am just gonna finish what I can on https://github.com/gratipay/gratipay.com/pull/4153 since that's anyway going to be needed for email signups. I'll try to find time this week for more. Good luck for https://github.com/gratipay/inside.gratipay.com/issues/757 :)

**Comment by whit537** [comment]
Thank you! :-)

**Comment by whit537** [comment]
> Also, I think it may be better to focus on bulk sign-up on `/on/npm/` instead of adding npm packages to search.

Todo updated to reflect this.

I just compiled my schedule for this week. I should have a little time tomorrow and Tuesday, and then I should be free from lunchtime onwards on Wednesday until I fall asleep. I would really rather not be trying to hack on Thursday morning before the lightning talk, so my goal would be to get everything done by Wednesday night.

Please forgive me, @aandis @rohitpaulk et al.: I may well cut lots of corners and self-merge PRs and whatnot in order to get this done. I really really want to demonstrate something real and interesting on Thursday! :rage4: Let's go!

**Comment by whit537** [comment]
I'm gonna pause on [decoupling](https://github.com/gratipay/inside.gratipay.com/issues/432) since that's pretty far along. The remaining changes to the About pages are low-risk. I'm going to focus here over the next couple days and come back to decoupling on Wednesday.

**Comment by mattbk** [comment]
> The remaining changes to the About pages are low-risk. 

If I get some time I'll dip into https://github.com/gratipay/gratipay.com/pull/4157 to move it along.

**Comment by whit537** [comment]
Okay! #4153 is deployed, and I've scheduled a daily job for 15:00 UTC.

![screen shot 2016-10-26 at 10 45 12 am](https://cloud.githubusercontent.com/assets/134455/19730845/7e42a6c4-9b69-11e6-8cbf-c95b4705d164.png)

**Comment by whit537** [comment]
Gonna get ready to head over to https://github.com/gratipay/inside.gratipay.com/issues/757, I'll check Papertrail before I leave. I won't be back for ~6 hours, at which point I plan to start hacking on the rest of this ticket.

**Comment by whit537** [comment]
> Your account has reached 100% log data transfer for this period. Log processing has been paused.

End of the month! Would you believe we got cut off, like two hours ago? 😞

**Comment by whit537** [comment]
Nothing yet. 😕 

```
gratipay::MAROON=> select count(*) from packages;
┌───────┐
│ count │
├───────┤
│     0 │
└───────┘
(1 row)
```

**Comment by whit537** [comment]
😕 

```
[gratipay] $ heroku logs -a gratipay | grep scheduler
[gratipay] $
```

**Comment by whit537** [comment]
"Last run" is now Oct 26 15:00 UTC. Maybe I will try it from `heroku run bash` ...

**Comment by whit537** [comment]
```
[gratipay] $ heroku run -a gratipay bin/sync-npm.sh > first-run.log 2>&1
```

```
[gratipay] $ tail -f first-run.log 
Running bin/sync-npm.sh on gratipay... starting, run.1126 (Hobby)
Running bin/sync-npm.sh on gratipay... connecting, run.1126 (Hobby)
Running bin/sync-npm.sh on gratipay... up, run.1126 (Hobby)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 24 27.2M   24 6784k    0     0   946k      0  0:00:29  0:00:07  0:00:22  825k
```

**Comment by whit537** [comment]
It's processing packages ...

```
econds
processed 2000 packages in   3 seconds
  0  201M    0 1167k    0     0   253k      0  0:13:32  0:00:04  0:13:28  253kprocessed 3000 packages in   4 seconds
  1  201M    1 2063k    0     0   367k      0  0:09:21  0:00:05  0:09:16  367kprocessed 4000 packages in   5 seconds
processed 5000 packages in   5 seconds
  1  201M    1 2895k    0     0   435k      0  0:07:52  0:00:06  0:07:46  503kprocessed 6000 packages in   6 seconds
processed 7000 packages in   6 seconds
  1  201M    1 3727k    0     0   490k      0  0:07:00  0:00:07  0:06:53  573kprocessed 8000 packages in   7 seconds
  2  201M    2 4431k    0     0   513k      0  0:06:41  0:00:08  0:06:33  809kprocessed 9000 packages in   8 seconds
  2  201M    2 5263k    0     0   549k      0  0:06:15  0:00:09  0:06:06  821kprocessed 10000 packages in   9 seconds
  2  201M    2 5967k    0     0   560k      0  0:06:07  0:00:10  0:05:57  776kprocessed 11000 packages in  10 seconds
  3  201M    3 6415k    0     0   549k      0  0:06:14  0:00:11  0:06:03  699kprocessed 12000 packages in  11 seconds
```

**Comment by whit537** [comment]
Oh heck yeah. Streaming! 💃

If I'm reading that right, it's sending data to postgres as it's discovered, not waiting to read the whole file.

**Comment by whit537** [comment]
Well, sorry. It's serializing as it downloads. I don't think it's uploading as it serializes.

**Comment by whit537** [comment]
Though some additional logging would help confirm that.

P.S. This play-by-play should maybe still be on #4153. :o)

**Comment by whit537** [comment]
Back from https://github.com/gratipay/gratipay.com/pull/4158. Rescheduling the sync ...

**Comment by whit537** [comment]
Scheduled job has kicked off!

**Comment by whit537** [comment]
```
2016-10-27T00:35:37.400471+00:00 heroku[scheduler.9462]: Process exited with status 0
```

**Comment by whit537** [comment]
Okay, we're live with the npm syncer.

```
gratipay::MAROON=> select count(*) from packages;
┌────────┐
│ count  │
├────────┤
│ 372271 │
└────────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
```
2016-10-27T00:35:26.231474+00:00 app[scheduler.9462]: processed 372271 packages in 247 seconds
2016-10-27T00:35:35.102567+00:00 heroku[scheduler.9462]: source=scheduler.9462 dyno=heroku.1942532.7efd96ec-cd0b-4718-b884-a779ee6f7de1 sample#load_avg_1m=0.77 sample#load_avg_5m=0.98
2016-10-27T00:35:35.102790+00:00 heroku[scheduler.9462]: source=scheduler.9462 dyno=heroku.1942532.7efd96ec-cd0b-4718-b884-a779ee6f7de1 sample#memory_total=140.20MB sample#memory_rss=34.25MB sample#memory_cache=105.94MB sample#memory_swap=0.00MB sample#memory_pgpgin=220943pages sample#memory_pgpgout=185564pages sample#memory_quota=512.00MB
2016-10-27T00:35:37.420115+00:00 heroku[scheduler.9462]: State changed from up to complete
2016-10-27T00:35:37.400471+00:00 heroku[scheduler.9462]: Process exited with status 0
```

**Comment by whit537** [comment]
Okay, so that looks like 247 seconds for the download/serialization, and another 11 seconds for the upsert. Call it ~260 seconds == 4.333 minutes.

**Comment by whit537** [comment]
We have a nightly backup from last night. I am taking another one just for safety.

**Comment by whit537** [comment]
Looks like `packages` adds 15 MB, or 21%.

![screen shot 2016-10-26 at 8 48 25 pm](https://cloud.githubusercontent.com/assets/134455/19750235/96299ad2-9bbd-11e6-932d-c8501c2cbd4b.png)

**Comment by whit537** [comment]
Curveball: some packages are unpublished without having a security package in their place. They show up in `all` and also on the registry, but [their JSON](http://registry.npmjs.com/zzzttt) is actually 404 and they have an unpublished time. They don't show up on site [search](https://www.npmjs.com/search?q=zzzttt) or [HTML](https://www.npmjs.com/package/zzzttt).

**Comment by whit537** [comment]
I'm adding the node buildpack back to our app so we can use `npm install marky-markdown`.

```
[gratipay] $ heroku run "npm install marky-markdown && sync-npm readmes"
Running npm install marky-markdown && sync-npm readmes on ⬢ gratipay... up, run.4576 (Hobby)
bash: npm: command not found
[gratipay] $ heroku buildpacks
=== gratipay Buildpack URL
https://github.com/gratipay/buildpack-python.git#gratipay-prod
[gratipay] $ heroku buildpacks:add heroku/nodejs
Buildpack added. Next release on gratipay will use:
  1. https://github.com/gratipay/buildpack-python.git#gratipay-prod
  2. heroku/nodejs
Run git push heroku master to create a new release using these buildpacks.
[gratipay] $ git push heroku master
```

**Comment by whit537** [comment]
```
gratipay::MAROON=> select count(*) from packages where readme_raw is not null;
┌───────┐
│ count │
├───────┤
│    22 │
└───────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
Okay! Gonna set this up in the scheduler.

**Comment by whit537** [comment]
Blorg. Hitting a memory limit. :-/

```
2016-10-27T03:30:45.084125+00:00 app[scheduler.6359]: svg2css
2016-10-27T03:30:45.088103+00:00 app[scheduler.6359]: splat-points-2d
2016-10-27T03:30:45.092127+00:00 app[scheduler.6359]: signeds3
2016-10-27T03:30:45.125973+00:00 app[scheduler.6359]: saltyrtc-task-webrtc
2016-10-27T03:30:45.127890+00:00 app[scheduler.6359]: resst-request
2016-10-27T03:30:45.723554+00:00 app[scheduler.6359]: 404 for zzzzzzzzzzzzzzzzzzz
2016-10-27T03:30:45.724534+00:00 app[scheduler.6359]: zzzttt
2016-10-27T03:30:45.979220+00:00 app[scheduler.6359]: 404 for zzzttt
2016-10-27T03:30:45.979404+00:00 app[scheduler.6359]: zzzss
2016-10-27T03:31:00.393328+00:00 heroku[scheduler.6359]: source=scheduler.6359 dyno=heroku.1942532.ee6fcf22-109a-49cf-aa0a-0136911a86ac sample#memory_total=641.52MB sample#memory_rss=103.62MB sample#memory_cache=4.98MB sample#memory_swap=532.91MB sample#memory_pgpgin=251526pages sample#memory_pgpgout=223723pages sample#memory_quota=512.00MB
2016-10-27T03:31:00.394012+00:00 heroku[scheduler.6359]: Process running mem=641M(124.3%)
2016-10-27T03:31:00.394066+00:00 heroku[scheduler.6359]: Error R14 (Memory quota exceeded)
```

**Comment by whit537** [comment]
Started again under scheduler ...

**Comment by whit537** [comment]
```
[gratipay] $ heroku logs --tail | grep scheduler
2016-10-27T03:50:20.064405+00:00 heroku[api]: Starting process with command `sync-npm readmes` by scheduler@ad
dons.heroku.com
2016-10-27T03:50:27.921066+00:00 heroku[scheduler.4775]: Starting process with command `sync-npm readmes`
2016-10-27T03:50:28.518982+00:00 heroku[scheduler.4775]: State changed from starting to up
2016-10-27T03:50:35.273088+00:00 heroku[scheduler.4775]: source=scheduler.4775 dyno=heroku.1942532.ea097600-5c
50-446b-bc32-d9ebbacfd652 sample#memory_total=35.95MB sample#memory_rss=31.01MB sample#memory_cache=4.95MB sam
ple#memory_swap=0.00MB sample#memory_pgpgin=14112pages sample#memory_pgpgout=5419pages sample#memory_quota=512
.00MB
2016-10-27T03:50:42.900770+00:00 app[scheduler.4775]: zzzzzzzzzzzzzzzzzzz
2016-10-27T03:50:42.933254+00:00 app[scheduler.4775]: utilise.clone
2016-10-27T03:50:42.936615+00:00 app[scheduler.4775]: svg2base64
2016-10-27T03:50:42.948038+00:00 app[scheduler.4775]: semver-bumper-for-file-text
2016-10-27T03:50:43.226081+00:00 app[scheduler.4775]: 404 for zzzzzzzzzzzzzzzzzzz
2016-10-27T03:50:43.226131+00:00 app[scheduler.4775]: zzzttt
2016-10-27T03:50:43.458823+00:00 app[scheduler.4775]: 404 for zzzttt
2016-10-27T03:50:43.458926+00:00 app[scheduler.4775]: zzzazzz
2016-10-27T03:50:44.033136+00:00 app[scheduler.4775]: 404 for zzzazzz
2016-10-27T03:50:44.033150+00:00 app[scheduler.4775]: zzz_012_censorify
2016-10-27T03:50:54.621207+00:00 app[scheduler.4775]: utilise.client
2016-10-27T03:50:55.719408+00:00 heroku[scheduler.4775]: source=scheduler.4775 dyno=heroku.1942532.ea097600-5c50-446b-bc32-d9ebbacfd652 sample#memory_total=360.68MB sample#memory_rss=355.62MB sample#memory_cache=5.00MB sample#memory_swap=0.06MB sample#memory_pgpgin=158661pages sample#memory_pgpgout=67876pages sample#memory_quota=512.00MB
2016-10-27T03:50:55.774905+00:00 app[scheduler.4775]: svg2android
2016-10-27T03:50:56.292132+00:00 app[scheduler.4775]: semver-bumper
```

**Comment by whit537** [comment]
```
gratipay::MAROON=> select count(*) from packages where readme_raw is not null;
┌───────┐
│ count │
├───────┤
│    82 │
└───────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
Call it 50 per minute. That's over five days. 😶

**Comment by whit537** [comment]
Alright, I'm about done here. We didn't make it.

Oh well. :-)

**Comment by whit537** [comment]
I've turned off both recurring tasks.

**Comment by whit537** [comment]
Sorry, I've removed them from the scheduler so that they don't recur. The existing process is still chugging along. It just crossed 1,000 readmes. We're coming up on 1100 at 20 minutes. Holding steady about 54 per minute, so I guess we let this run for five days?

**Comment by kaguillera** [comment]
I don't think so @mattbk. If I am not mistaken @whit537 the five days is for the initial loading and processing of the npm readme. The future updates would be less since it should only update the ones that have changed and add new ones that should not be that much.

**Comment by kaguillera** [comment]
!m @whit537

**Comment by mattbk** [comment]
Thank goodness.

**Comment by whit537** [comment]
Yes, though I had forgotten that the Heroku scheduler kills processes when the time limit is up, so I am just restarting it now.

**Comment by kaguillera** [comment]
😞

**Comment by whit537** [comment]
```
from packages where readme_raw is not null) / (select count(*) from packages)::float;
┌───────────────────┐
│     ?column?      │
├───────────────────┤
│ 0.114843756295817 │
└───────────────────┘
(1 row)

gratipay::MAROON=> 






  7 remote                         
```

**Comment by whit537** [comment]
26%

**Comment by whit537** [comment]
35%

I count 109,354 unique email addresses.

```sql
select unnest(emails) as email, count(name) from packages group by email;
```

**Comment by whit537** [comment]
| npackages | nemails |
|---------:|-----------:|
| 1,000+ | 8 |
| 100-999 | 462 |
| 10-99 | 10,266 |
| 2-9 | 52,567 |
| 1 | 46,051 |

**Comment by whit537** [comment]
Looks like npm is stripping a leading H1.

![screen shot 2016-10-28 at 4 16 29 pm](https://cloud.githubusercontent.com/assets/134455/19820981/ed704576-9d29-11e6-86de-b88174fe583f.png)

![screen shot 2016-10-28 at 4 16 51 pm](https://cloud.githubusercontent.com/assets/134455/19820989/f6b6f06c-9d29-11e6-974a-a3d731a3c135.png)

**Comment by whit537** [comment]
Hmm ... https://www.npmjs.com/package/marky-markdown#npm-packages

**Comment by whit537** [comment]
Syncers are running again! https://github.com/gratipay/inside.gratipay.com/issues/884#issuecomment-259320916

I guess when we upsert we'll want to go off the time from back when we first started loading things up a couple weeks ago, back at https://github.com/gratipay/gratipay.com/issues/4148#issuecomment-256370122, to make sure we pick up any changes since then.

**Comment by whit537** [comment]
Fetcher crashed!

```
2016-11-09T02:45:47.778173+00:00 app[scheduler.7320]: Traceback (most recent call last):
2016-11-09T02:45:47.779195+00:00 app[scheduler.7320]:   File "/app/.heroku/python/bin/sync-npm", line 9, in <module>
2016-11-09T02:45:47.779275+00:00 app[scheduler.7320]:     load_entry_point('gratipay', 'console_scripts', 'sync-npm')()
2016-11-09T02:45:47.779327+00:00 app[scheduler.7320]:   File "/app/gratipay/package_managers/sync.py", line 160, in main
2016-11-09T02:45:47.779601+00:00 app[scheduler.7320]:     globals()[args.command.replace('-', '_')](env, args, db)
2016-11-09T02:45:47.779619+00:00 app[scheduler.7320]:   File "/app/gratipay/package_managers/sync.py", line 140, in fetch_readmes
2016-11-09T02:45:47.779835+00:00 app[scheduler.7320]:     k-component-page_readmes.fetch(db)
2016-11-09T02:45:47.779868+00:00 app[scheduler.7320]:
2016-11-09T02:45:47.779887+00:00 app[scheduler.7320]:   File "/app/gratipay/package_managers/readmes.py", line 93, in fetch
2016-11-09T02:45:47.781538+00:00 app[scheduler.7320]:     threaded_map(Fetcher(db), dirty, 4)
2016-11-09T02:45:47.781590+00:00 app[scheduler.7320]:   File "/app/gratipay/utils/threaded_map.py", line 23, in threaded_map
2016-11-09T02:45:47.781767+00:00 app[scheduler.7320]:     raise e.args[0]
2016-11-09T02:45:47.782241+00:00 app[scheduler.7320]: psycopg2.ProgrammingError: can't adapt type 'dict'
```

**Comment by whit537** [comment]
Meanwhile the Processor is starting.

**Comment by whit537** [comment]
Also crashed!

```
2016-11-09T02:50:45.011994+00:00 app[scheduler.7918]: Traceback (most recent call last):
2016-11-09T02:50:45.012249+00:00 app[scheduler.7918]:   File "/app/.heroku/python/bin/sync-npm", line 9, in <module>
2016-11-09T02:50:45.012320+00:00 app[scheduler.7918]:     load_entry_point('gratipay', 'console_scripts', 'sync-npm')()
2016-11-09T02:50:45.012355+00:00 app[scheduler.7918]:   File "/app/gratipay/package_managers/sync.py", line 160, in main
2016-11-09T02:50:45.012543+00:00 app[scheduler.7918]:     globals()[args.command.replace('-', '_')](env, args, db)
2016-11-09T02:50:45.012573+00:00 app[scheduler.7918]:   File "/app/gratipay/package_managers/sync.py", line 144, in process_readmes
2016-11-09T02:50:45.012635+00:00 app[scheduler.7918]:     _readmes.process(db)
2016-11-09T02:50:45.012663+00:00 app[scheduler.7918]:   File "/app/gratipay/package_managers/readmes.py", line 97, in process
2016-11-09T02:50:45.012887+00:00 app[scheduler.7918]:     dirty = db.all('SELECT id, package_manager, name, description, readme_raw '
2016-11-09T02:50:45.012917+00:00 app[scheduler.7918]:   File "/app/.heroku/python/lib/python2.7/site-packages/postgres/__init__.py", line 548, in all
2016-11-09T02:50:45.013917+00:00 app[scheduler.7918]:     return cursor.all(sql, parameters)
2016-11-09T02:50:45.013947+00:00 app[scheduler.7918]:   File "/app/.heroku/python/lib/python2.7/site-packages/postgres/cursors.py", line 145, in all
2016-11-09T02:50:45.014009+00:00 app[scheduler.7918]:     self.execute(sql, parameters)
2016-11-09T02:50:45.014036+00:00 app[scheduler.7918]:   File "/app/.heroku/python/lib/python2.7/site-packages/psycopg2/extras.py", line 288, in execute
2016-11-09T02:50:45.014475+00:00 app[scheduler.7918]:     return super(NamedTupleCursor, self).execute(query, vars)
2016-11-09T02:50:45.014555+00:00 app[scheduler.7918]: psycopg2.ProgrammingError: syntax error at or near "BY"
2016-11-09T02:50:45.014557+00:00 app[scheduler.7918]: LINE 1: ... packages WHERE readme_needs_to_be_processedORDER BY package...
```

**Comment by whit537** [comment]
I'm surprised the test suite didn't catch that one. Hmm ...

**Comment by whit537** [comment]
Ah! Now I see that we weren't testing that code. 😊

**Comment by whit537** [comment]
PR in https://github.com/gratipay/gratipay.com/pull/4176 ...

**Comment by whit537** [comment]
Hmm ... readme processor is running out of memory.

```
$ heroku logs --tail | grep scheduler
2016-11-09T03:50:28.311155+00:00 heroku[api]: Starting process with command `sync-npm process-readmes` by scheduler@addons.heroku.com
2016-11-09T03:50:35.164998+00:00 heroku[scheduler.8487]: Starting process with command `sync-npm process-readmes`
2016-11-09T03:50:35.846204+00:00 heroku[scheduler.8487]: State changed from starting to up
2016-11-09T03:50:47.749213+00:00 heroku[scheduler.8487]: source=scheduler.8487 dyno=heroku.1942532.1fbd95bf-81f1-4a26-aa9f-a30e88765924 sample#memory_total=142.36MB sample#memory_rss=137.41MB sample#memory_cache=4.95MB sample#memory_swap=0.00MB sample#memory_pgpgin=41624pages sample#memory_pgpgout=5691pages sample#memory_quota=512.00MB
2016-11-09T03:51:07.948771+00:00 heroku[scheduler.8487]: source=scheduler.8487 dyno=heroku.1942532.1fbd95bf-81f1-4a26-aa9f-a30e88765924 sample#memory_total=371.86MB sample#memory_rss=366.91MB sample#memory_cache=4.95MB sample#memory_swap=0.00MB sample#memory_pgpgin=100375pages sample#memory_pgpgout=5691pages sample#memory_quota=512.00MB
2016-11-09T03:51:28.261222+00:00 heroku[scheduler.8487]: source=scheduler.8487 dyno=heroku.1942532.1fbd95bf-81f1-4a26-aa9f-a30e88765924 sample#load_avg_1m=0.29
2016-11-09T03:51:28.261321+00:00 heroku[scheduler.8487]: source=scheduler.8487 dyno=heroku.1942532.1fbd95bf-81f1-4a26-aa9f-a30e88765924 sample#memory_total=1107.29MB sample#memory_rss=494.66MB sample#memory_cache=0.00MB sample#memory_swap=612.63MB sample#memory_pgpgin=406200pages sample#memory_pgpgout=280077pages sample#memory_quota=512.00MB
2016-11-09T03:51:28.261959+00:00 heroku[scheduler.8487]: Process running mem=1107M(216.3%)
2016-11-09T03:51:28.261959+00:00 heroku[scheduler.8487]: Error R15 (Memory quota vastly exceeded)
2016-11-09T03:51:28.262044+00:00 heroku[scheduler.8487]: Stopping process with SIGKILL
2016-11-09T03:51:28.473986+00:00 heroku[scheduler.8487]: Process exited with status 137
2016-11-09T03:51:28.519921+00:00 heroku[scheduler.8487]: State changed from up to complete
```

**Comment by whit537** [comment]
It's never even getting to the point where it emits a log line about processing even a single readme.

**Comment by whit537** [comment]
Ah, I bet it's because we're including `readme_raw` in the initial query over all unprocessed packages.

**Comment by whit537** [comment]
PR in https://github.com/gratipay/gratipay.com/pull/4177.

**Comment by whit537** [comment]
Blorg.

```
Traceback (most recent call last):
  File "/app/.heroku/python/bin/sync-npm", line 9, in <module>
    load_entry_point('gratipay', 'console_scripts', 'sync-npm')()
  File "/app/gratipay/package_managers/sync.py", line 160, in main
    globals()[args.command.replace('-', '_')](env, args, db)
  File "/app/gratipay/package_managers/sync.py", line 144, in process_readmes
    _readmes.process(db)
  File "/app/gratipay/package_managers/readmes.py", line 100, in process
    threaded_map(Processor(db), dirty, 4)
  File "/app/gratipay/utils/threaded_map.py", line 23, in threaded_map
    raise e.args[0]
OSError: net.js:639
    throw new TypeError('invalid data');
    ^

TypeError: invalid data
  at Socket.write (net.js:639:11)
  at /app/bin/our-marky-markdown.js:20:18
  at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:404:3)
```

**Comment by whit537** [comment]
Unable to reproduce locally. Too jetlagged, more tomorrow ...

**Comment by whit537** [comment]
Lessee here ...

**Comment by whit537** [comment]
I can repro the bug using a one-off dyno.

**Comment by whit537** [comment]
Still works locally. Hmm ...

**Comment by whit537** [comment]
https://github.com/gratipay/gratipay.com/issues/4148#issuecomment-259330483 points to net.js and fs.js. Let's see if we can get a better error log ...

**Comment by whit537** [comment]
```
~ $ sync-npm process-readmes
zzzzzzzzzzzzzzzzzzz
utilise.emitterify
svg2png-cli
semver-max
zzzzzzxl
react-router-native
plane-to-polygon
Traceback (most recent call last):
  File "/app/gratipay/utils/threaded_map.py", line 15, in g
    return func(*a, **kw)
  File "/app/gratipay/package_managers/readmes.py", line 72, in process
    processed = markdown.render_like_npm(raw)
  File "/app/gratipay/utils/markdown.py", line 44, in render_like_npm
    raise OSError(err)
OSError: net.js:639
    throw new TypeError('invalid data');
    ^

TypeError: invalid data
  at Socket.write (net.js:639:11)
  at /app/bin/our-marky-markdown.js:20:18
  at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:404:3)



Traceback (most recent call last):
  File "/app/.heroku/python/bin/sync-npm", line 9, in <module>
    load_entry_point('gratipay', 'console_scripts', 'sync-npm')()
  File "/app/gratipay/package_managers/sync.py", line 160, in main
    globals()[args.command.replace('-', '_')](env, args, db)
  File "/app/gratipay/package_managers/sync.py", line 144, in process_readmes
    _readmes.process(db)
  File "/app/gratipay/package_managers/readmes.py", line 100, in process
    threaded_map(Processor(db), dirty, 4)
  File "/app/gratipay/utils/threaded_map.py", line 23, in threaded_map
    raise e.args[0]
OSError: net.js:639
    throw new TypeError('invalid data');
    ^

TypeError: invalid data
  at Socket.write (net.js:639:11)
  at /app/bin/our-marky-markdown.js:20:18
  at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:404:3)


planet-news
planet-names
planet-maps-raster-2
planet-maps-raster
planet-maps
planet-icons
planet-hacker
planet-forty
planet-feeds
planetfall-archive
planet-facts
planet-express
planet-digest
planet-css
planet-client
planet-classic
planet-blank
planetary-system
planetary.js
planetary
planet
planer
planepacker
planemo
plane-mesh
```

**Comment by whit537** [comment]
Is it a problem with stdout?

**Comment by whit537** [comment]
> The 1st argument given to [`.write()`](http://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback) is expected to be a `String` or `Buffer`.

http://stackoverflow.com/a/26851940

**Comment by whit537** [comment]
What is `marky` returning that is not a `String` or `Buffer`?

**Comment by whit537** [comment]
I made a local modification to just process packages with names starting with `pla`, but wasn't able to trigger the error that way.

**Comment by whit537** [comment]
I overwrote the `our-marky-markdown.js` script on a Heroku dyno (using here strings) to give some debugging output.

`html for nullisn't a string, it's a functionfunction (selector, context, r, opts)`

**Comment by whit537** [comment]
Alright, so we need to call `.html()` on the output of `marky()`. But why doesn't it fail locally? I have node 4.3.1 where Heroku has 5.11.1. Maybe the `write` API changed?

**Comment by whit537** [comment]
[Nope](https://nodejs.org/dist/v4.3.1/docs/api/stream.html#stream_writable_write_chunk_encoding_callback).

**Comment by whit537** [comment]
Maybe I don't have the right data loaded up locally?

**Comment by whit537** [comment]
Well, they do differ.

# Local

```
gratipay=# select count(*) from packages;
┌────────┐
│ count  │
├────────┤
│ 372147 │
└────────┘
(1 row)

gratipay=# select count(*) from packages where readme_raw is not null;
┌───────┐
│ count │
├───────┤
│   134 │
└───────┘
(1 row)

gratipay=#
```

# Remote

```
gratipay::MAROON=> select count(*) from packages;                                                             ┌────────┐
│ count  │
├────────┤
│ 372271 │
└────────┘
(1 row)

gratipay::MAROON=> select count(*) from packages where readme_raw is not null;
┌────────┐
│ count  │
├────────┤
│ 138260 │
└────────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
Oh! I'm seeing marky-markdown v9.0.0 locally, but 8.1.0 on Heroku.

**Comment by whit537** [comment]
Cached build?

**Comment by whit537** [comment]
For the fetcher bug (https://github.com/gratipay/gratipay.com/issues/4148#issuecomment-259322332), I think we'll need to unthread in order to get better error messaging.

**Comment by whit537** [comment]
Actually, `threaded_map` is [supposed](https://github.com/gratipay/gratipay.com/blob/2001/gratipay/utils/threaded_map.py#L22) to output the original traceback to stdout. Are we losing that? I guess we'll have to reschedule to see ...

**Comment by whit537** [comment]
Well that is consternating. Even after https://github.com/gratipay/gratipay.com/pull/4178 I am still seeing 8.1.0 on a Heroku dyno. Hmm ...

**Comment by whit537** [comment]
[Got it](https://github.com/gratipay/inside.gratipay.com/issues/884#issuecomment-259559065)!

**Comment by whit537** [comment]
`sync-npm process-readmes` is working from a one-off dyno. 👍

**Comment by whit537** [comment]
I'm not seeing an error from `sync-npm fetch-readmes` in a one-off dyno. I've rescheduled both that and `process-readmes`.

**Comment by whit537** [comment]
Alright, the original traceback is in there, I had just confused it for a dupe.

```
2016-11-09T23:51:23.098776+00:00 app[scheduler.8465]: jstuningTraceback (most recent call last):
2016-11-09T23:51:23.098779+00:00 app[scheduler.8465]:   File "/app/gratipay/utils/threaded_map.py", line 15, i
n g
2016-11-09T23:51:23.098780+00:00 app[scheduler.8465]:     return func(*a, **kw)
2016-11-09T23:51:23.098781+00:00 app[scheduler.8465]:   File "/app/gratipay/package_managers/readmes.py", line
 55, in fetch
2016-11-09T23:51:23.098781+00:00 app[scheduler.8465]:     , dirty.name
2016-11-09T23:51:23.098782+00:00 app[scheduler.8465]:   File "/app/.heroku/python/lib/python2.7/site-packages/
postgres/__init__.py", line 374, in run
2016-11-09T23:51:23.098783+00:00 app[scheduler.8465]:     cursor.run(sql, parameters)
2016-11-09T23:51:23.098784+00:00 app[scheduler.8465]:   File "/app/.heroku/python/lib/python2.7/site-packages/
postgres/cursors.py", line 92, in run
2016-11-09T23:51:23.098784+00:00 app[scheduler.8465]:     self.execute(sql, parameters)
2016-11-09T23:51:23.098785+00:00 app[scheduler.8465]:   File "/app/.heroku/python/lib/python2.7/site-packages/
psycopg2/extras.py", line 288, in execute
2016-11-09T23:51:23.098786+00:00 app[scheduler.8465]:     return super(NamedTupleCursor, self).execute(query, 
vars)
```

**Comment by whit537** [comment]
Interesting. So maybe `name` or something can be a dict coming from npm?

**Comment by whit537** [comment]
That doesn't sound right ...

**Comment by whit537** [comment]
I'm working on a PR to add Sentry support to these procs so we have better visibility into and resiliency in the face of errors (surely these won't be the last).

**Comment by whit537** [comment]
PR in #4179.

**Comment by whit537** [comment]
So it turns out that the registry includes packages that have been unpublished—a lot of them, from what I can tell. They appear as JSON with a 404. E.g. below.

Currently we log "404" for these and leave `readme_raw` untouched, which means it'll still be null next time around and we'll refetch the same 404. We should notice this case and probably drop the record from our database, though we'll want to be careful to bring it back again when someone else claims it.

I'm a little surprised at how many 404s I'm seeing. How does unpublishing relate to deleting or removing a package?

http://registry.npmjs.com/mysql-schema
https://www.npmjs.com/package/mysql-schema

```json
{  
    "_id":"mysql-schema",
    "_rev":"4-c527f6a64778f8d0afbaf6fd4754085e",
    "name":"mysql-schema",
    "time":{  
        "modified":"2013-12-08T03:17:01.297Z",
        "created":"2013-12-08T03:16:59.506Z",
        "0.0.1":"2013-12-08T03:17:01.297Z",
        "unpublished":{  
            "name":"carlosmarte",
            "time":"2014-07-26T06:27:28.217Z",
            "tags":{  
                "latest":"0.0.1"
            },
            "maintainers":[  
                {  
                    "name":"carlosmarte",
                    "email":"dev@carlosmarte.me"
                }
            ],
            "description":"mysql queries helper",
            "versions":[  
                "0.0.1"
            ]
        }
    },
    "_attachments":{  

    }
}
```

**Comment by whit537** [comment]
Okay! Fetcher and processor appear to be chugging along. We want to see these both hit zero, though they won't until we delete or skip 404s.

```
gratipay::MAROON=> select count(*) from packages where readme_raw is null;
┌────────┐
│ count  │
├────────┤
│ 223221 │
└────────┘
(1 row)

gratipay::MAROON=> select count(*) from packages where readme_needs_to_be_processed;
┌────────┐
│ count  │
├────────┤
│ 323828 │
└────────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
Done in https://github.com/gratipay/gratipay.com/pull/4181 and deployed.

**Comment by whit537** [comment]
Deleting! 👍 

```
2016-11-10T07:50:24.124188+00:00 app[scheduler.4899]: fetching practiceone
2016-11-10T07:50:23.863759+00:00 app[scheduler.9696]: fetching node-u2f
2016-11-10T07:50:23.939573+00:00 app[scheduler.9696]: fetching node-typograph
2016-11-10T07:50:23.974219+00:00 app[scheduler.9696]: no readme in killdrev
2016-11-10T07:50:23.974273+00:00 app[scheduler.9696]: fetching kill-desktop-osx
2016-11-10T07:50:24.020858+00:00 app[scheduler.9696]: fetching infinigon-tag
2016-11-10T07:50:24.027898+00:00 app[scheduler.9696]: fetching kill-dash-nine
2016-11-10T07:50:24.037882+00:00 app[scheduler.9696]: 404 for spurious-js-aws-sdk-helper
2016-11-10T07:50:24.038176+00:00 app[scheduler.9696]: fetching spur-di
2016-11-10T07:50:24.081055+00:00 app[scheduler.9696]: fetching node-typo
2016-11-10T07:50:24.152517+00:00 app[scheduler.4899]: yet-another-module is 404; deleting
2016-11-10T07:50:24.155182+00:00 app[scheduler.4899]: fetching yet-another-friendly-dependency
2016-11-10T07:50:24.185761+00:00 app[scheduler.4899]: fetching practice-npm-package
2016-11-10T07:50:24.203793+00:00 app[scheduler.4899]: fetching jspm-nodelibs-process
2016-11-10T07:50:24.228972+00:00 app[scheduler.4899]: fetching jspm-nodelibs-path
2016-11-10T07:50:24.273577+00:00 app[scheduler.4899]: fetching hwsl2
2016-11-10T07:50:24.279907+00:00 app[scheduler.4899]: fetching jspm-nodelibs-os
2016-11-10T07:50:24.304598+00:00 app[scheduler.4899]: yet-another-friendly-dependency is 404; deleting
2016-11-10T07:50:24.308818+00:00 app[scheduler.4899]: fetching yet-another-express-routing
2016-11-10T07:50:24.348231+00:00 app[scheduler.4899]: fetching practice_npm
2016-11-10T07:50:24.358175+00:00 app[scheduler.4899]: fetching jspm-nodelibs-net
2016-11-10T07:50:24.367534+00:00 app[scheduler.4899]: yet-another-express-routing is 404; deleting
2016-11-10T07:50:24.370328+00:00 app[scheduler.4899]: fetching yet-another-express-router
2016-11-10T07:50:24.132520+00:00 app[scheduler.9696]: fetching node-typhoon
2016-11-10T07:50:24.162294+00:00 app[scheduler.9696]: 404 for spur-di
2016-11-10T07:50:24.162538+00:00 app[scheduler.9696]: fetching sptitesmith-stylus-retina-template
2016-11-10T07:50:24.182196+00:00 app[scheduler.9696]: fetching kill-combo
2016-11-10T07:50:24.218516+00:00 app[scheduler.9696]: 404 for sptitesmith-stylus-retina-template
2016-11-10T07:50:24.218923+00:00 app[scheduler.9696]: fetching spruce
2016-11-10T07:50:24.253262+00:00 app[scheduler.9696]: no readme in spruce
2016-11-10T07:50:24.253270+00:00 app[scheduler.9696]: fetching sprout-object
2016-11-10T07:50:24.268288+00:00 app[scheduler.9696]: fetching node-typewriter
2016-11-10T07:50:24.291538+00:00 app[scheduler.9696]: 404 for sprout-object
```

**Comment by whit537** [comment]
I'm curious to see how many records we're left with after weeding out the 404s.

Okay! I'm gonna let this run overnight ...

**Comment by whit537** [comment]
Almost done fetching! Processing is slower to catch up ...

```
gratipay::MAROON=> select count(*) from packages where readme_raw is null;
┌───────┐
│ count │
├───────┤
│  3349 │
└───────┘
(1 row)

gratipay::MAROON=> select count(*) from packages where readme_needs_to_be_processed;
┌────────┐
│ count  │
├────────┤
│ 258387 │
└────────┘
(1 row)

gratipay::MAROON=> select count(*) from packages;
┌────────┐
│ count  │
├────────┤
│ 345044 │
└────────┘
(1 row)
```

**Comment by whit537** [comment]
No movement on first and third. I think we have some packages that aren't 404 but also maybe don't have a readme? I think that's how readme_raw is ending up null for a percentage.

```
gratipay::MAROON=> select count(*) from packages where readme_raw is null;
┌───────┐
│ count │
├───────┤
│  3349 │
└───────┘
(1 row)

gratipay::MAROON=> select count(*) from packages where readme_needs_to_be_processed;
┌────────┐
│ count  │
├────────┤
│ 230868 │
└────────┘
(1 row)

gratipay::MAROON=> select count(*) from packages;
┌────────┐
│ count  │
├────────┤
│ 345044 │
└────────┘
(1 row)
```

**Comment by whit537** [comment]
27,519 readmes processed in five hours, call it 5,000 an hour, so ... 45-50 hours remaining? Should be done over the weekend?

**Comment by whit537** [comment]
230,868 - 207,714 = 23,154 in about five hours. 👍

**Comment by whit537** [comment]
> Fetcher crashed!

Captured with #4179: https://sentry.io/gratipay/gratipay-com/issues/179430708/. 👍

**Comment by whit537** [comment]
`mysql-prettify` has a readme of `{"private": true}`. It [shows](https://www.npmjs.com/package/mysql-prettify) as a string on the HTML npm; looking at the [json](http://registry.npmjs.com/mysql-prettify) I do see an object there (though my browser is not prettifying it for some reason? ironic given the name 😛 ).

**Comment by whit537** [comment]
I don't see a repo for `mysql-prettify`. I was gonna check `package.json` to see if that's where the bad value is coming from. Do we want to stringify it or count is as "no readme"?

**Comment by whit537** [comment]
Alright, the `{"private": true}` issue should be fixed in #4182.

**Comment by whit537** [comment]
I'm going to fix the other Sentry bug that I introduced: https://sentry.io/gratipay/gratipay-com/issues/179441800/.

**Comment by whit537** [comment]
PR in #4183. Waiting for Travis.

```
gratipay::MAROON=> select count(*) from packages where readme_raw is null;
┌───────┐
│ count │
├───────┤
│     2 │
└───────┘
(1 row)

gratipay::MAROON=> select count(*) from packages where readme_needs_to_be_processed;
┌───────┐
│ count │
├───────┤
│ 97562 │
└───────┘
(1 row)

gratipay::MAROON=> select count(*) from packages;
┌────────┐
│ count  │
├────────┤
│ 345044 │
└────────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
207,714 - 97,476 = 110,238 in about 19 hours = 5,802. On track! 👍

**Comment by whit537** [comment]
```
gratipay::MAROON=> select count(*) from packages where readme_raw is not null;
┌────────┐
│ count  │
├────────┤
│ 345042 │
└────────┘
(1 row)

gratipay::MAROON=> select count(*) from packages where readme_needs_to_be_processed;
┌───────┐
│ count │
├───────┤
│     3 │
└───────┘
(1 row)

gratipay::MAROON=> select count(*) from packages;
┌────────┐
│ count  │
├────────┤
│ 345044 │
└────────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
Here are the three remaining to be processed:

https://www.npmjs.com/package/testing233
https://www.npmjs.com/package/testing234
https://www.npmjs.com/package/kendo-ui-react-jquery-stockchart

The first two give "ERROR: No README data found!" The third results in a 504!

**Comment by nobodxbodon** [comment]
Is this part of a plan to build an index of all the popular libraries? Next is Maven rep for Java, if this works out?

**Comment by whit537** [comment]
Yeah, something like that. Ideally we can partner with Libraries.io and bring a bunch online at once.

**Comment by whit537** [comment]
¯\\\_(ツ)\_/¯

```
gratipay::MAROON=> select count(*) from packages where readme_raw is not null;
┌────────┐
│ count  │
├────────┤
│ 345044 │
└────────┘
(1 row)

gratipay::MAROON=> select count(*) from packages where readme_needs_to_be_processed;
┌───────┐
│ count │
├───────┤
│     0 │
└───────┘
(1 row)

gratipay::MAROON=> select count(*) from packages;
┌────────┐
│ count  │
├────────┤
│ 345044 │
└────────┘
(1 row)

gratipay::MAROON=>
```

**Comment by whit537** [comment]
Alright! READMEs initially loaded! 💃

Let's get some pages on the site! #4151

**Comment by whit537** [comment]
Slight change of plans. #4151 is too much of a rabbit hole. We don't to get ourselves into the business of processing and securing READMEs across 30+ package managers. Our existing `/on/network/foo/` pages aren't very contentful, there's no reason `/on/npm/foo/` need be, either. My current plan is to make a PR to remove the README processing that is already deployed (we still want package fetching and syncing, since emails are in there and that's our key for linking with users), and then move on to Checkpoint 2: Giving to Packages.

**Comment by whit537** [comment]
@kaguillera and I are talking IRL about how much tech debt we want to take on underneath [Relax Open Work Requirement](https://github.com/orgs/gratipay/projects/5?fullscreen=true) in order to fast-track the npm feature here. Over there, we are renaming Teams to projects and members to collaborators, and we are also now [talking](https://github.com/gratipay/gratipay.com/pull/4219) about removing tip migration. It's basically a question of whether we change names in the UI only, or also remove code and drop/rename database tables and such. The trade-off is that if we only make surface changes over there, then the next @JessaWitzel that comes along will have _even more_ confusion to deal with ("Wait—projects are stored in the teams table? WTF!")—but I am going to be _so incredibly mad_ if we're not first to market with npm pledging. :rage4: 

@JessaWitzel Can we please please please go into debt here? I PROMISE I will fix it and make it all better in January or February. Or March. 🙏

**Comment by whit537** [comment]
Discussing in [slack](http://gratipay.slackarchive.io/gratipay/-/1480696458.000793/1480708090.000895/1480708035000887/).

**Comment by aandis** [comment]
> January or February. Or March. 

😄

**Comment by whit537** [comment]
> Ok. I will bless this technical debt with my fairy wand but request a deadline for fixing it.
> 12 weeks after feature launch

@JessaWitzel at [slack](http://gratipay.slackarchive.io/gratipay/-/1480707779.000875/1480709421.000978/1480709359000968/)

**Comment by whit537** [comment]
We have reached **Checkpoint 1: Inert /on/npm/foo/ Pages**! 💃 

https://gratipay.com/on/npm/react-router/

![screen shot 2016-12-04 at 7 13 48 pm](https://cloud.githubusercontent.com/assets/134455/20870416/ce65cd06-ba55-11e6-9d26-ab8bd1b00a53.png)

**Comment by nobodxbodon** [comment]
Anything I can do to accelarate this? It seems top priority for now.

One question: is NPM cool with this, and are they willing to partner/coordinate in any way?

**Comment by whit537** [comment]
Thanks for bumping this, @nobodxbodon! As mentioned in [slack](http://gratipay.slackarchive.io/gratipay/-/1483308579.001479/1483477112.00159/1483469265001564/), I hoping to spec this out this week while also bringing [Relax Open Work Requirement](https://github.com/orgs/gratipay/projects/5?fullscreen=true) in for landing (this was blocked on that).

**Comment by whit537** [comment]
> is NPM cool with this, and are they willing to partner/coordinate in any way?

I emailed them and didn't hear back. I think if we get some traction with this, that will be the time to reapproach a conversation with them.

> any workload estimate and roadmap with timeline?

Last month I guesstimated ([one](http://gratipay.slackarchive.io/gratipay/-/1480704365.000861/1480709302.000964/1480709162000959/), [two](http://gratipay.slackarchive.io/gratipay/-/1481837154.000682/1482002861.000787/1481923658000760/)) that the [two](https://github.com/orgs/gratipay/projects/5?fullscreen=true) [projects](https://github.com/orgs/gratipay/projects/8?fullscreen=true) together would take six weeks of calendar time. It's now been four and a half weeks and we're not done with either yet. We're likely to finish the first by the end of this week (Week 5), with next week being Week 1 on [Integrate npm](https://github.com/orgs/gratipay/projects/8?fullscreen=true). It seems unlikely to take a week. ;-) I was off by a factor of 2.5 on the [Relax Open Work](https://github.com/orgs/gratipay/projects/5?fullscreen=true) time estimate (I figured two out of the six). That suggests 10 weeks of calendar time for Integrate npm.

Roadmap and further estimation tbd when I can spec this out.

> who are main developers and any dividing of tasks?

Me and maybe @aandis? Anybody else wanna volunteer? :-) Division of tasks tbd.

> any blockers?

https://github.com/orgs/gratipay/projects/5?fullscreen=true

**Comment by whit537** [comment]
I've put both projects on the [calendar](http://inside.gratipay.com/appendices/calendar), with an initial target date of March 17 for Integrate npm.

**Comment by whit537** [comment]
Honestly this is probably going to go into April or May, but we can take that as it comes. :-)

**Comment by aandis** [comment]
I'd really like to see this happening on gratipay so I am ready to help as much as I can. We need a development plan so that things are moving along.

**Comment by mattbk** [comment]
> I've put both projects on the calendar

This would be more useful in Gantt chart format, IMO. The calendar is cool for events, but in terms of lining up work, it's just a bunch of blue bars to me.

**Comment by whit537** [comment]
[Conversation](http://gratipay.slackarchive.io/gratipay/-/1484261894.002684/1484586804.002789/1484586154002766/) about product developement and wire-framing, might try https://gomockingbird.com/.

**Comment by whit537** [comment]
Closing per https://github.com/gratipay/inside.gratipay.com/issues/987#issuecomment-272985143.
